File: src/android/com/bluvision/beacons/plugin/BluProvisionWrapper.java
package com.bluvision.beacons.plugin;

import com.bluvision.sdk.beacons.Beacon;
import com.bluvision.sdk.beacons.Blufi;
import com.bluvision.sdk.beacons.SBeacon;
import com.bluvision.sdk.cloud.Error;
import com.bluvision.sdk.cloud.Template;
import com.bluvision.sdk.cloud.User;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import java.math.BigInteger;
import java.util.Date;
import java.util.List;

/**
 * This class echoes a string called from JavaScript.
 */
public class BluProvisionWrapper extends CordovaPlugin implements BeaconInteractorCallback{

    private BeaconInteractor mBeaconInteractor;
    private CallbackContext scanCallbackContext;
    private CallbackContext connectionCallbackContext;
    private CallbackContext templateCallbackContext;
    private String deviceType = "SBeacon";

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("init")) {
            deviceType = args.getString(0);
            this.init(callbackContext);
            return true;
        } else if (action.equals("signIn")) {
            String token = args.getString(0);
            this.signIn(callbackContext, token);
            return true;
        } else if (action.equals("signOut")) {
            this.signOut();
            return true;
        } else if (action.equals("getTemplate")) {
            String beacon = args.getString(0);
            this.getTemplates(callbackContext, beacon);
            return true;
        } else if (action.equals("provision")) {
            JSONObject data = args.getJSONObject(0);
            String beacon = data.getString("id");
            String beaconName = data.getString("name");
            String notes = data.getString("notes");
            int template = data.getInt("templateId");
            if(notes == null){
              notes = "";
            }
            this.provisionBeacon(callbackContext, beacon, template, beaconName, notes);
            return true;
        } else if (action.equals("startScan")) {
            this.startScan(callbackContext);
            return true;
        } else if (action.equals("stopScan")) {
            this.stopScan();
            return true;
        }
        return false;
    }

    private void init(CallbackContext callbackContext) {
        Context context = this.cordova.getActivity().getApplicationContext();
        Activity activity = this.cordova.getActivity();
        mBeaconInteractor = new BeaconInteractor();
        mBeaconInteractor.init(context, this, activity);
    }

    private void provisionBeacon(CallbackContext callbackContext, String beacon, int template, String beaconName, String notes) {
        connectionCallbackContext = callbackContext;
        mBeaconInteractor.provisionBeacon(beacon, template, beaconName, notes);
    }
    private void getTemplates(CallbackContext callbackContext, String id) {
        templateCallbackContext = callbackContext;
        mBeaconInteractor.loadTemplates(id);
    }

    private void signIn(CallbackContext callbackContext, String token) {
        scanCallbackContext = callbackContext;
        mBeaconInteractor.signIn(token);
    }

    private void signOut() {
        mBeaconInteractor.signOut();
        stopScan();
    }

    private void startScan(CallbackContext callbackContext) {
      scanCallbackContext = callbackContext;
      mBeaconInteractor.startScan();
    }

    private void stopScan() {
        scanCallbackContext = null;
        mBeaconInteractor.stopScan();
    }

    @Override
    public void beaconFound(Beacon beacon){
        JSONObject result = new JSONObject();
        try {
            if(deviceType.equals(beacon.getTypeString())){
              result.put("code",  "Beacon_Found");
              result.put("name",  beacon.getBluetoothDevice().getName());
              if (beacon instanceof SBeacon) {
                BigInteger sid64 = ((SBeacon) beacon).getIdentifier();
                String identifier = sid64.toString();
                String hexIdentifier = sid64.toString(16).toUpperCase();
                result.put("id", identifier);
                result.put("hex", hexIdentifier);
                result.put("address",  beacon.getBluetoothDevice().getAddress());
                result.put("type",  beacon.getTypeString());
                result.put("rssi",  beacon.getRSSI());
              }
              if (beacon instanceof Blufi){
                result.put("type",  beacon.getTypeString());
              }
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.OK, result);
         pr.setKeepCallback(true);
         if (this.scanCallbackContext != null) {
             this.scanCallbackContext.sendPluginResult(pr);
         }
    }

    @Override
    public void beaconLost(Beacon beacon) {
        JSONObject result = new JSONObject();
        try {
          if(deviceType.equals(beacon.getTypeString())){
            result.put("code",  "Beacon_Lost");
            result.put("name",  beacon.getBluetoothDevice().getName());
            if (beacon instanceof SBeacon) {
                BigInteger sid64 = ((SBeacon) beacon).getIdentifier();
                String identifier = sid64.toString();
                String hexIdentifier = sid64.toString(16).toUpperCase();
                result.put("id", identifier);
                result.put("hex", hexIdentifier);
                result.put("address",  beacon.getBluetoothDevice().getAddress());
                result.put("type",  beacon.getTypeString());
            }
            if (beacon instanceof Blufi){
                result.put("type",  beacon.getTypeString());
            }
          }
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.OK, result);
         pr.setKeepCallback(true);
         if (this.scanCallbackContext != null) {
             this.scanCallbackContext.sendPluginResult(pr);
         }
    }

    @Override
    public void signInSuccess(User user) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("code", "SignIn_Success");
            jsonObject.put("user_name", user.getUsername());
            jsonObject.put("user_email", user.getEmail());
            jsonObject.put("project_id", user.getProjectId());
            jsonObject.put("project_name", user.getProjectName());
            jsonObject.put("current_project", user.getCurrentProject());
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.OK, jsonObject);
         pr.setKeepCallback(true);
         if (this.scanCallbackContext != null) {
             this.scanCallbackContext.sendPluginResult(pr);
         }
    }

    @Override
    public void signInFailure(Error error) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("Fail", "SignIn_Failed");
            jsonObject.put("error_code", error.getCode());
            jsonObject.put("error_message", error.getMessage());
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.ERROR, jsonObject);
         pr.setKeepCallback(false);
         if (this.scanCallbackContext != null) {
             this.scanCallbackContext.sendPluginResult(pr);
             scanCallbackContext = null;
         }
    }

    @Override
    public void enableBluetooth() {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("Fail", "Bluetooth");
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.ERROR, jsonObject);
         pr.setKeepCallback(false);
         if (this.scanCallbackContext != null) {
             this.scanCallbackContext.sendPluginResult(pr);
             scanCallbackContext = null;
         }
    }

    @Override
    public void requestLocationPermissions() {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("Fail", "Location");
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.ERROR, jsonObject);
         pr.setKeepCallback(false);
         if (this.scanCallbackContext != null) {
             this.scanCallbackContext.sendPluginResult(pr);
             scanCallbackContext = null;
         }
    }

    @Override public void loadTemplateSucess(List<? extends Template> list) {
      JSONObject resultJson = new JSONObject();
        try {
            JSONArray jsonArray = new JSONArray();
            for(Template template: list) {
                JSONObject jsonObject = new JSONObject();
                jsonObject.put("templateId", template.getTemplateId());
                jsonObject.put("name", template.getName());
                jsonArray.put(jsonObject);
            }
            resultJson.put("templates", jsonArray);
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.OK, resultJson);
         pr.setKeepCallback(false);
         if (this.templateCallbackContext != null) {
             this.templateCallbackContext.sendPluginResult(pr);
             templateCallbackContext = null;
         }
    }

    @Override public void loadTemplateError(String s) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("Template_Fail", s);
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.ERROR, jsonObject);
         pr.setKeepCallback(false);
         if (this.templateCallbackContext != null) {
             this.templateCallbackContext.sendPluginResult(pr);
             templateCallbackContext = null;
         }
    }

    @Override public void provisionSuccess() {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("code", "Provision_Sucess");
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.OK, jsonObject);
         pr.setKeepCallback(false);
         if (this.connectionCallbackContext != null) {
             this.connectionCallbackContext.sendPluginResult(pr);
             connectionCallbackContext = null;
         }
    }

    @Override public void provisionError(String message) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("result", "ERROR");
            jsonObject.put("reason", message);
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.ERROR, jsonObject);
         pr.setKeepCallback(false);
         if (this.connectionCallbackContext != null) {
             this.connectionCallbackContext.sendPluginResult(pr);
             connectionCallbackContext = null;
         }
    }

    @Override public void provisionStatusChange(String s) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("code", "Status_Changed");
            jsonObject.put("result", s);
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.OK, jsonObject);
         pr.setKeepCallback(true);
         if (this.connectionCallbackContext != null) {
             this.connectionCallbackContext.sendPluginResult(pr);
         }
    }

    @Override public void provisiondescriptionChange(String s) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("code", "Desc_Changed");
            jsonObject.put("result", s);
        } catch (JSONException e) {
            e.printStackTrace();
        }
         PluginResult pr = new PluginResult(Status.OK, jsonObject);
         pr.setKeepCallback(true);
         if (this.connectionCallbackContext != null) {
             this.connectionCallbackContext.sendPluginResult(pr);
         }
    }
}

File: src/ios/com/bluvision/beacons/plugin/BluProvisionWrapper.m
/********* BluProvisionWrapper.m Cordova Plugin Implementation *******/

#import <Cordova/CDV.h>
#import <Bluvision/BLUBluFi.h>
#import "BeaconInteractor.h"

@interface BluProvisionWrapper : CDVPlugin <BeaconInteractorDelegate, CBCentralManagerDelegate> {
    BeaconInteractor *interactor;
    NSString *deviceType;
    CDVInvokedUrlCommand *locationCommand;
    CDVInvokedUrlCommand *bluCommand;
    CDVInvokedUrlCommand *scanCommand;
    CDVInvokedUrlCommand *connectioncommand;
    CDVInvokedUrlCommand *templateCommand;
}

@property (nonatomic, strong) CLLocationManager *locationManager;
@property (nonatomic, strong) CBCentralManager *bluManager;
@property (nonatomic, strong) NSTimer *timer;

- (void)bluetoothPermission:(CDVInvokedUrlCommand*)command;
- (void)locationPermission:(CDVInvokedUrlCommand*)command;
- (void)init:(CDVInvokedUrlCommand*)command;
- (void)signIn:(CDVInvokedUrlCommand*)command;
- (void)startScan:(CDVInvokedUrlCommand*)command;
- (void)stopScan:(CDVInvokedUrlCommand*)command;
- (void)getTemplate:(CDVInvokedUrlCommand*)command;
- (void)provision:(CDVInvokedUrlCommand*)command;

@end

@implementation BluProvisionWrapper

- (void)bluetoothPermission:(CDVInvokedUrlCommand*)command {

    bluCommand = command;
    [self blutoothPermission];
}

- (void)locationPermission:(CDVInvokedUrlCommand*)command {

    locationCommand = command;
    [self requestAlwaysAuth];
}

- (void)requestAlwaysAuth {

    CLAuthorizationStatus status = [CLLocationManager authorizationStatus];

    if (status == kCLAuthorizationStatusNotDetermined) {

        self.locationManager = [[CLLocationManager alloc] init];
        [self.locationManager requestWhenInUseAuthorization];
        self.timer = [NSTimer scheduledTimerWithTimeInterval:1.0
                                                      target:self
                                                    selector:@selector(checkStatus)
                                                    userInfo:nil repeats:YES];
    } else if (status == kCLAuthorizationStatusAuthorizedWhenInUse || status == kCLAuthorizationStatusAuthorizedAlways) {

        // send success
        NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
        [dictionary setValue:@"Location_Success" forKey:@"code"];
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
        [pluginResult setKeepCallbackAsBool:NO];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:locationCommand.callbackId];
    } else if (status == kCLAuthorizationStatusDenied) {

        [self requestLocationPermissions];
    }
}

- (void)blutoothPermission {

    self.bluManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil];
}

- (void)centralManagerDidUpdateState:(CBCentralManager *)central {

    if (self.bluManager.state == CBManagerStatePoweredOn) {

        NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
        [dictionary setValue:@"Blutooth_Success" forKey:@"code"];
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
        [pluginResult setKeepCallbackAsBool:NO];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:bluCommand.callbackId];
    } else {

        NSString *stateString = @"";
        if (self.bluManager.state == CBCentralManagerStateResetting) {
            stateString = @"The connection with the system service was momentarily lost, update imminent.";
        } else if (self.bluManager.state == CBCentralManagerStateUnsupported) {
            stateString = @"The platform doesn't support Bluetooth Low Energy.";
        } else if (self.bluManager.state == CBCentralManagerStateUnauthorized) {
            stateString = @"The app is not authorized to use Bluetooth Low Energy.";
        } else if (self.bluManager.state == CBCentralManagerStatePoweredOff) {
            stateString = @"Bluetooth is currently powered off.";
        } else if (self.bluManager.state == CBManagerStateUnknown) {
            stateString = @"State unknown, update imminent.";
        }
        [self enableBluetooth:stateString];
    }
}

- (void)checkStatus {

    CLAuthorizationStatus status = [CLLocationManager authorizationStatus];

    if (status == kCLAuthorizationStatusAuthorizedWhenInUse || status == kCLAuthorizationStatusAuthorizedAlways) {

        [self.timer invalidate];

        NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
        [dictionary setValue:@"Location_Success" forKey:@"code"];
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
        [pluginResult setKeepCallbackAsBool:NO];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:locationCommand.callbackId];
    } else {
        if (status==kCLAuthorizationStatusDenied) {
            // denied
            [self.timer invalidate];

            [self requestLocationPermissions];
        }
    }
}

- (void)init:(CDVInvokedUrlCommand*)command {

    scanCommand = command;
    NSString* device = [command.arguments objectAtIndex:0];
    if (device != nil && device.length > 0) {
        deviceType = device;
    } else {
        deviceType = @"SBeacon";
    }
    interactor = [[BeaconInteractor alloc] initWithDelegate:self];
}

- (void)signIn:(CDVInvokedUrlCommand*)command {

    scanCommand = command;
    NSString* token = [command.arguments objectAtIndex:0];
    if (token != nil) {
        [interactor signIn:token device:deviceType];
    }
}

- (void)startScan:(CDVInvokedUrlCommand*)command {

    scanCommand = command;
    [interactor startScan:deviceType];
}

- (void)stopScan:(CDVInvokedUrlCommand*)command {

    scanCommand = nil;
    [interactor stopScan];
}

- (void)getTemplate:(CDVInvokedUrlCommand*)command {

    templateCommand = command;
    NSString* beacondentifier = [command.arguments objectAtIndex:0];
    if (beacondentifier != nil) {
        [interactor loadTemplates:beacondentifier];
    }
}

- (void)provision:(CDVInvokedUrlCommand*)command {

    connectioncommand = command;
    NSDictionary* dictionary = [command.arguments objectAtIndex:0];
    NSString* template = [dictionary valueForKey:@"templateId"];
    NSString* notes = [dictionary valueForKey:@"notes"];
    NSString* name = [dictionary valueForKey:@"name"];
    if (notes == nil) {
        notes = @"";
    }
    if (template != nil) {

        [interactor provisionBeaconForTemplate:[template intValue] beaconName:name notes:notes];
    }
}

#pragma mark - BeaconInteractorDelegate

- (void)signInSuccess:(BZCUser *)user {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"SignIn_Success" forKey:@"code"];
    [dictionary setValue:user.username forKey:@"user_name"];
    [dictionary setValue:user.email forKey:@"user_email"];
    [dictionary setValue:user.currentProject.identifier forKey:@"project_id"];
    [dictionary setValue:user.currentProject.name forKey:@"project_name"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:scanCommand.callbackId];
}

- (void)signInFailure:(NSError *)error {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"SignIn_Failed" forKey:@"Fail"];
    [dictionary setValue:[NSString stringWithFormat:@"%ld", (long)error.code] forKey:@"error_code"];
    [dictionary setValue:error.localizedDescription forKey:@"error_message"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:NO];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:scanCommand.callbackId];
    scanCommand = nil;
}

- (void)beaconFound:(BLUSBeacon *)beacon {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"Beacon_Found" forKey:@"code"];
    [dictionary setValue:beacon.name forKey:@"name"];
    if ([beacon isKindOfClass:[BLUBluFi class]]) {
        [dictionary setValue:deviceType forKey:@"type"];
        [dictionary setValue:beacon.identifier.stringValue forKey:@"id"];
    } else {
        [dictionary setValue:beacon.identifier.stringValue forKey:@"id"];
        [dictionary setValue:[self convertToHex:beacon.identifier.stringValue].uppercaseString forKey:@"hex"];
        [dictionary setValue:beacon.device.macAddress forKey:@"address"];
        [dictionary setValue:deviceType forKey:@"type"];
    }

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:scanCommand.callbackId];
}

- (void)beaconLost:(BLUSBeacon *)beacon {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"Beacon_Lost" forKey:@"code"];
    [dictionary setValue:beacon.name forKey:@"name"];
    [dictionary setValue:beacon.RSSI forKey:@"rssi"];
    if ([beacon isKindOfClass:[BLUBluFi class]]) {
        [dictionary setValue:beacon.device.typeString forKey:@"type"];
    } else {
        [dictionary setValue:beacon.identifier.stringValue forKey:@"id"];
        [dictionary setValue:[self convertToHex:beacon.identifier.stringValue].uppercaseString forKey:@"hex"];
        [dictionary setValue:beacon.device.macAddress forKey:@"address"];
        [dictionary setValue:beacon.device.typeString forKey:@"type"];
    }

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:scanCommand.callbackId];
}

- (void)enableBluetooth:(NSString *)message {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:message forKey:@"result"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:NO];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:bluCommand.callbackId];
    bluCommand = nil;
}

- (void)requestLocationPermissions {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"Location permission required" forKey:@"result"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:NO];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:locationCommand.callbackId];
    locationCommand = nil;
}

- (void)loadTemplateSucess:(NSArray *)templates {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];

    NSMutableArray *array = [[NSMutableArray alloc] init];
    for (BZCTemplate *template in templates) {

        NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
        [dict setValue:template.identifier.stringValue forKey:@"templateId"];
        [dict setValue:template.name forKey:@"name"];
        [array addObject:dict];
    }

    [dictionary setValue:array forKey:@"templates"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:NO];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:templateCommand.callbackId];
}

- (void)loadTemplateError:(NSString *)message {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:message forKey:@"Template_Fail"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:NO];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:templateCommand.callbackId];
    templateCommand = nil;
}

- (void)provisionSuccess {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"Provision_Sucess" forKey:@"code"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:NO];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:connectioncommand.callbackId];
}

- (void)provisionError:(NSString *)message {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"ERROR" forKey:@"result"];
    [dictionary setValue:message forKey:@"reason"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:NO];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:connectioncommand.callbackId];
    connectioncommand = nil;
}

- (void)provisionStatusChange:(NSString *)status {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"Status_Changed" forKey:@"code"];
    [dictionary setValue:status forKey:@"result"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:connectioncommand.callbackId];
}

- (void)provisionDescriptionChange:(NSString *)description {

    NSMutableDictionary *dictionary = [[NSMutableDictionary alloc] init];
    [dictionary setValue:@"Desc_Changed" forKey:@"code"];
    [dictionary setValue:description forKey:@"result"];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: dictionary];
    [pluginResult setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:connectioncommand.callbackId];
}

#pragma mark - Private Method

- (NSString *)convertToHex: (NSString *)string {

    NSString *hexString = [NSString stringWithFormat:@"%@",
                           [NSData dataWithBytes:[string cStringUsingEncoding:NSUTF8StringEncoding]
                                          length:strlen([string cStringUsingEncoding:NSUTF8StringEncoding])]];

    for(NSString * toRemove in [NSArray arrayWithObjects:@"<", @">", @" ", nil])
        hexString = [string stringByReplacingOccurrencesOfString:toRemove withString:@""];

    return hexString;
}

@end
